### PR TITLE
[None][fix] Warm up flashinfer sampling module during engine warmup

### PR DESCRIPTION
**Pull request by qiaoxj07**
## Description

flashinfer ships its sampling kernels as source and JIT-builds them with `nvcc` on first use. The only thing that exercises the non-greedy sampler during warmup today is the advanced-sampling CUDA graph capture pass (`_run_capture_pass(force_non_greedy=True)`), so with `cuda_graph_config: null` nothing triggers that build until the first non-greedy request is actually served.

The build then runs **inside the executor loop** and takes ~80–105s with a cold JIT cache. The cache lives under `$HOME` inside the container (`/root/.cache/flashinfer/...`), so it is cold on every job. While one rank compiles, it never launches its MoE all-to-all dispatch, so peer ranks GPU-spin on its completion flag and block in `tp_gather`. If the build crosses the 300s hang-detector threshold, the detector fires and aborts the whole job.

This was observed on a 16-node GB300 disaggregated DeepSeek-V4 run. The hang detector caught a CTX rank with this stack:

```
py_executor.py    _executor_loop_overlap -> _forward_step
model_engine.py   model_forward
sampler/ops/flashinfer.py:349  sample_from_logits_op
sampler/ops/flashinfer.py:232  top_k_mask_logits_op
flashinfer/sampling.py:1977    top_k_mask_logits
flashinfer/sampling.py:68      get_sampling_module
flashinfer/jit/core.py:316     build_and_load
flashinfer/jit/cpp_ext.py:368  run_ninja      <-- blocked in nvcc >300s
```

while its peers sat in `_flush_pending_transfer_responses -> _enqueue_responses -> tp_gather`.

Generation workers were unaffected in the same runs because they set `cuda_graph_config.batch_sizes`, which makes the advanced-sampling capture pass run and build the module during warmup.

## Fix

Build the module during engine warmup instead, where the hang detector is not armed and no peer is waiting on this rank. `get_sampling_module()` covers every sampling kernel the sampler uses and is `@functools.cache`d upstream, so this is a no-op once built.

Not gated on spec-decoding: the plain `TorchSampler` non-greedy path reaches the same kernels, so gating would leave that case exposed. Cost is one build at startup that would otherwise be paid on the first non-greedy request anyway.

## Test Coverage

- Verified the no-flashinfer branch returns cleanly before any `flashinfer.*` attribute access (the module binds a placeholder that raises `ImportError` on attribute access when flashinfer is absent).
- Verified idempotency: repeated calls build once.
- **Cluster validation** on two 4-node GB300 disagg runs (ctx `cuda_graph_config: null` + MTP), patch bind-mounted over the container's modules:

  |  | unpatched (prod, 16-node) | patched A | patched B |
  |---|---|---|---|
  | iterations reached | **died at iter 9** | **2255** | **2287** |
  | mid-serving stall #1 | **98.2s** | 20.0s | 19.2s |
  | mid-serving stall #2 | **58.1s** | — | 10.6s |
  | `Hang detected` events | yes → `MPI_Abort` | **0** | **0** |

  The `ninja: Entering .../cached_ops/sampling` block now appears during warmup, ahead of the first warmup forward, and never during serving. Patched run B additionally carried `TORCH_LOGS=recompiles`: 111 recompiles were logged, all in the DeepSeek-V4 sparse-attention helpers, and none inside any stall window — so Dynamo/Inductor is not a contributor.

  Not a controlled A/B: validation runs are 4-node `ctx1 + gen1` @ concurrency 224 vs. 16-node `ctx4 + gen1` @ 1418 for the failures. See the [validation comment](https://github.com/NVIDIA/TensorRT-LLM/pull/17286#issuecomment-5189122881) for the full breakdown.

## PR Checklist

- [x] Commit is signed off (DCO)
- [x] PR title follows `[JIRA/NVBUG/None][type] description`
- [x] One concern per PR

## Dev Engineer Review

- Added `warmup_sampling_module()` to eagerly initialize cached FlashInfer sampling kernels.
- Calls the helper at the start of `PyTorchModelEngine.warmup()`, before early-return guards and other warmup stages.
- Makes prewarming best-effort by logging initialization failures without aborting startup.
- Preserves the no-FlashInfer path.
- Prevents first-use NVCC JIT compilation in the executor loop for non-greedy requests.
- No public API declarations, configuration files, or test-list files changed.
- Validation covered missing FlashInfer handling, idempotency, warmup ordering, and GB300 disaggregated cluster startup behavior.
- Reported CI runs failed and require investigation and another run.
- The cluster comparison was not controlled because the configurations and scales differed.

## QA Engineer Review

No test changes.

### CODE REVIEW

**Comment by qiaoxj07** [comment]
Addressed the CodeRabbit finding in 3546e88.

The call is now hoisted to the top of `warmup()`, above **both** early returns — the flagged non-HELIX context-parallel guard *and* the `kv_cache_manager is None` guard, which is the same class of problem. Neither guard has any bearing on whether the sampling kernels are needed, and `warmup_sampling_module()` reads no engine state, so it is safe on every path.

### Cluster validation

Running now on a 4-node GB300 disagg config that reproduces the preconditions (ctx `cuda_graph_config: null` + MTP `max_draft_len: 3`, gen with pinned `cuda_graph_config.batch_sizes`). The patched files are bind-mounted over the container's own modules.

Early signal is what the fix predicts: the `ninja: Entering .../cached_ops/sampling` blocks now appear **while the server is still starting up** (`5_wait_server` at 240s elapsed, not yet healthy), instead of mid-serving. Will post the final iteration timings once it reaches steady state.

For reference, this is the stack the hang detector caught on an unpatched run, which is what motivated the change:

```
py_executor.py                 _executor_loop_overlap -> _forward_step
model_engine.py                model_forward
sampler/ops/flashinfer.py:349  sample_from_logits_op
sampler/ops/flashinfer.py:232  top_k_mask_logits_op
flashinfer/sampling.py:68      get_sampling_module
flashinfer/jit/core.py:316     build_and_load
flashinfer/jit/cpp_ext.py:368  run_ninja      <-- blocked in nvcc >300s
```

File: tensorrt_llm/_torch/pyexecutor/sampler/ops/flashinfer.py [hunk 1]
```diff
@@ -353,6 +353,19 @@ def sample_from_logits_op(
...
+def warmup_sampling_module() -> None:
+    """Build flashinfer's sampling kernels now instead of on first use.
+
+    flashinfer ships these kernels as source, so the first sampling call runs
+    nvcc inline (~90s with a cold cache). Doing that inside the executor loop
+    stalls the rank past the hang detector's threshold. Cheap no-op once built.
+    """
+    if not IS_FLASHINFER_AVAILABLE:
+        return
+    flashinfer.sampling.get_sampling_module()
```
Comment (zhaoyangwang-nvidia) [inline tensorrt_llm/_torch/pyexecutor/sampler/ops/flashinfer.py]: **nit:** worth a `try/except Exception` + `logger.warning` here — now that this runs above every guard in `warmup()`, a broken JIT toolchain aborts startup even for greedy-only deployments that never touch these kernels, whereas both neighbouring pre-JIT warmups are deliberately best-effort.

**Comment by qiaoxj07** [comment]
## Cluster validation complete

Two 4-node GB300 disagg runs with this patch bind-mounted over the container's modules, on a config that reproduces the preconditions (ctx `cuda_graph_config: null` + MTP `max_draft_len: 3`).

### Mid-serving stalls

|  | unpatched (prod) | patched A | patched B |
|---|---|---|---|
| iterations reached | **died at iter 9** | **2255** | **2287** |
| stall #1 | **98.2s** (iter 6) | 20.0s (iter 6) | 19.2s (iter 6) |
| stall #2 | **58.1s** (iter 7) | — | 10.6s (iter 512) |
| fatal | `Hang detected` → `MPI_Abort` | **0** | **0** |
| how it ended | hang detector | hit `job_time` | hit `job_time` |

~156s of multi-minute mid-serving stalls collapse to ~20s, and both patched runs served 2250+ iterations with zero hang-detector events. The unpatched run died at iteration 9.

The `ninja: Entering .../cached_ops/sampling` block appears during warmup in both runs, ahead of the first warmup forward, and never again during serving.

### torch.compile ruled out

Patched run B also carried `TORCH_LOGS=recompiles`. It logged 111 recompiles, all in the DeepSeek-V4 sparse-attention helpers (`_compute_compressed_mask`, `_compute_gen_compressed_position_ids`, `_prepare_deepseek_v4_indices_compiled`, `_compute_per_ratio_kv_lens`). **None fall inside any stall window** — the windows are silent. So Dynamo/Inductor recompilation is not a contributor here.

### Caveat

This is not a controlled A/B. The validation runs are 4-node `ctx1 + gen1` at concurrency 224; the failing production runs are 16-node `ctx4 + gen1` at concurrency 1418. What is demonstrated is that the patched configuration serves 2250+ iterations without the multi-minute stalls or the hang. A production-scale rerun would be needed to claim the failure rate itself is fixed.

**Comment by qiaoxj07** [comment]
Good catch — applied in 25611d3.

You're right that hoisting the call above the guards changed the blast radius: before, a broken JIT toolchain only surfaced at the first non-greedy request (and never at all for greedy-only deployments); after, it aborted startup for everyone. The prewarm is an optimization, so it should never be able to do that.

Now wrapped in `try` / `logger.warning`, and the docstring says why. Deployments that actually sample non-greedily still hit the same error at the real call site, so nothing is silently swallowed that matters.

On the exception breadth: I went with a broad catch rather than matching the DeepGEMM prewarm's narrow `except RuntimeError`. flashinfer's build can fail as a `subprocess.CalledProcessError` re-raised as `RuntimeError` (`run_ninja`), a `MissingJITCacheError` when `FLASHINFER_DISABLE_JIT` is set, or an `OSError` from `tvm_ffi.load_module`. The DeepGEMM comment justifies its narrowness by wanting signature drift to surface loudly — that doesn't apply here, since `get_sampling_module()` takes no arguments.
